homeassistant/components/group/climate.py

```diff
@@ -0,0 +1,390 @@
+            x.state
+            for x in states
+            if x.state not in [HVACMode.OFF, STATE_UNAVAILABLE, STATE_UNKNOWN]
+        ]
+        # return the most common hvac mode (what the thermostat is set to do) except OFF, UNKNOWN and UNAVAILABE
+        if current_hvac_modes:
+            self._attr_hvac_mode = HVACMode(
+                max(sorted(set(current_hvac_modes)), key=current_hvac_modes.count)
+            )
+        # return off if any is off
+        elif any(x.state == HVACMode.OFF for x in states):
+            self._attr_hvac_mode = HVACMode.OFF
+        # else it's none
+        else:
+            self._attr_hvac_mode = None
+        # return the most common action if it is not off
+        hvac_actions = list(find_state_attributes(states, ATTR_HVAC_ACTION))
+        current_hvac_actions = [a for a in hvac_actions if a != HVACAction.OFF]
+        # return the most common action if it is not off
+        if current_hvac_actions:
+            self._attr_hvac_action = max(
+                sorted(set(current_hvac_actions)), key=current_hvac_actions.count
+            )
+        # return action off if all are off
+        elif all(a == HVACAction.OFF for a in hvac_actions):
+            self._attr_hvac_action = HVACAction.OFF
+        # else it's none
+        else:
+            self._attr_hvac_action = None
+
+        # available swing modes
+        all_swing_modes = list(find_state_attributes(states, ATTR_SWING_MODES))
+        self._attr_swing_modes = merge_modes(all_swing_modes)
+
+        # Report the most common swing_mode.
+        self._attr_swing_mode = most_frequent_attribute(states, ATTR_SWING_MODE)
+
+        # available fan modes
+        all_fan_modes = list(find_state_attributes(states, ATTR_FAN_MODES))
+        self._attr_fan_modes = merge_modes(all_fan_modes)
+
+        # Report the most common fan_mode.
+        self._attr_fan_mode = most_frequent_attribute(states, ATTR_FAN_MODE)
+
+        # available preset modes
+        all_preset_modes = list(find_state_attributes(states, ATTR_PRESET_MODES))
+        self._attr_preset_modes = merge_modes(all_preset_modes)
+
+        # Report the most common fan_mode.
+        self._attr_preset_mode = most_frequent_attribute(states, ATTR_PRESET_MODE)
+
+        # Supported flags
+        self._attr_supported_features = ClimateEntityFeature(0)
+        for support in find_state_attributes(states, ATTR_SUPPORTED_FEATURES):
+            # Merge supported features by emulating support for every feature
+            # we find.
+            self._attr_supported_features |= support
+
+        # Bitwise-and the supported features with the Grouped climate's features
+        # so that we don't break in the future when a new feature is added.
+        all_supported_features = ClimateEntityFeature(0)
+        for feature_flags in self._features:
+            all_supported_features |= feature_flags
+        self._attr_supported_features &= all_supported_features
```

Comment:
Why not adding new features if available by input entities?

Reply:
The set of features present in the `_features` dictionary are the ones that this implementation is supporting (by implementing corresponding methods from ClimateEntity). We don't want to report the group to support a feature (that could be added in the future) without explicitly updating this group implementation to support it. This logic therefore restricts that the supported features are the intersection between the ones from the input entities and the ones the group class has implemented.

Reply:
Why not adding it to `self._features` instead to allow dynamically expanding with new features.
That was what I meant with one of my very first comments on `self._features` that why isn't it constructed dynamically for everything possible from `ClimateEntityFeature`

Reply:
Because with the definition of the keys in a dynamic way, we can hit in the future that a new feature is added, and when a feature is added a method also needs to be implemented with its corresponding logic. This can happen that in the future some one else adds a new feature, this gets dynamically added as a key to the `_features` dict, and then when called the service to the ClimateGroup that is related to that feature, we would get some error about not being implemented. Therefore is safer to be explicit about which are the features that the `ClimateGroup` is supporting rather than expect it to be dynamically defined without any changes on the logic.

Reply:
It was not setting the variable twice but computing the intersection between the features from the entities and the ones supported by the ClimateGroup class.

Indeed the logic could be reduced to fewer lines of code so I did this.